### PR TITLE
Fix: `validate_input_col` for partial functions

### LINKED ISSUE #95066: `validate_input_col` is impractical for partial functions (with large arguments)
### 🐛 Describe the bug

One can map a `Callable` over datapipes using [torchdata.datapipes.iter.Mapper](https://pytorch.org/data/main/generated/torchdata.datapipes.iter.Mapper.html#torchdata.datapipes.iter.Mapper) where `Callable` might be a `functools.partial`.

The mapper ultimately [calls](https://github.com/pytorch/pytorch/blob/2f9ffe7b0ae61c804d0aec898c4ac3946c109e41/torch/utils/data/datapipes/iter/callable.py#L84) `validate_input_col` ([def](https://github.com/pytorch/pytorch/blob/2f9ffe7b0ae61c804d0aec898c4ac3946c109e41/torch/utils/data/datapipes/utils/common.py#L24)) which in turn calls `str()` on the `Callable` to get its name.

* `str(Callable)` is not the correct way to get the `fn_name`
* `str()` on a `functools.partial` that has large arguments is very slow
* `str()` on a `functools.partial` includes the complete function arguments in the returned string. These will then be printed in a potential stacktrace...

```python3
import functools
from timeit import default_timer as timer

from torch.utils.data.datapipes.utils.common import validate_input_col


def foo(*args):
    pass

d = {i: list(range(i)) for i in range(10_000)}
partial_foo = functools.partial(foo, d)

start = timer()
validate_input_col(fn=partial_foo, input_col=[1, 2])
end = timer()
print(f"elapsed time: {round(end - start, 2)}")
```
-> `elapsed time: 6.21`

changing `def foo(*args):` to `def foo(arg1)` results in a stacktrace with around 300 million characters, because it includes the whole `d`.

### Versions

> PyTorch version: 1.13.1+cu117
Is debug build: False
Python version: 3.9.12 (main, Feb  1 2023, 14:04:09)  [GCC 9.4.0] (64-bit runtime)

cc @VitalyFedyunin @ejguan @NivekT @dzhulgakov

### PR DESCRIPTION
**Pull request by BeckerFelix**
Fixes #95066

#### Proposed change:
do not call `str()` on a `Callable` to determine its name

#### Reasoning:
Please see https://github.com/pytorch/pytorch/issues/95066 for reasoning and examples

#### Effect:
* The code example given in https://github.com/pytorch/pytorch/issues/95066 now executes instantly.
* If invalid input is provided, the stacktrace now prints nicely as
  ```
  ValueError: The function foo takes 1 parameters, but 2 are required.
  ```

### CODE REVIEW

**Review by NivekT — APPROVED** [review]
LGTM! Thanks for spotting this issue and fixing it. I don't think this should cause any issue for lambda function right?

cc: @ejguan I will cherry-pick this into the release.

**Comment by BeckerFelix** [comment]
> LGTM! Thanks for spotting this issue and fixing it. I don't think this should cause any issue for lambda function right?
> 
> cc: @ejguan I will cherry-pick this into the release.

thanks for the review. 
`__name__` is valid on lambda functions, so we should be good.

**Review by ejguan — CHANGES_REQUESTED** [review]
Can you pls add tests to `test/test_datapipe.py`?
You can inject them into `test_map_tuple_list_with_col_iterdatapipe` and `test_map_dict_with_col_iterdatapipe`

**Comment by BeckerFelix** [comment]
> test_map_tuple_list_with_col_iterdatapipe

Happy to enhance the tests. Can you be more specific of what exactly you would like to see tested, please? In both `test_map_tuple_list_with_col_iterdatapipe` and `test_map_dict_with_col_iterdatapipe` map functions, partials, and lambdas get mapped over a datapipe. 
I could provide a test which creates a partial with a large input argument and maps this over the datapipe. Similar to the code snippet I provided in https://github.com/pytorch/pytorch/issues/95066. wdyt?

**Comment by ejguan** [comment]
> I could provide a test which creates a partial with a large input argument and maps this over the datapipe. Similar to the code snippet I provided in #95066. wdyt?

Sounds good to me.

**Comment by BeckerFelix** [comment]
> 

let me know what you think about d4b3ed997ae603379ca93798c87744ecde69c5e1, please

**Review by ejguan — APPROVED** [review]

File: torch/utils/data/datapipes/utils/common.py [hunk 1]
```diff
@@ -78,6 +77,11 @@ def validate_input_col(fn: Callable, input_col: Optional[Union[int, tuple, list]
         else:
             continue
 
+    if isinstance(fn, functools.partial):
+        fn_name = fn.func.__name__
+    else:
+        fn_name = fn.__name__
```
Comment (ejguan) [inline torch/utils/data/datapipes/utils/common.py:83]: This causes lots of problem as not all of callable objects have `__name__`
I am going to do a forward fix.

Comment (ejguan) [reply torch/utils/data/datapipes/utils/common.py:83]: Simple example would be a `nn.Module`

**Pull request closed without merging**